pkg/api/aim2/dao/repositories/experiment.go

```diff
@@ -44,149 +52,247 @@ func NewExperimentRepository(db *gorm.DB) *ExperimentRepository {
-	if err := r.db.Transaction(func(tx *gorm.DB) error {
+	return r.db.Transaction(func(tx *gorm.DB) error {
 		if err := tx.WithContext(ctx).Model(&experiment).Updates(experiment).Error; err != nil {
 			return eris.Wrapf(err, "error updating experiment with id: %d", *experiment.ID)
 		}
 
 		// also archive experiment runs if experiment is being archived
 		if experiment.LifecycleStage == models.LifecycleStageDeleted {
-			run := models.Run{
-				LifecycleStage: experiment.LifecycleStage,
-				DeletedTime:    experiment.LastUpdateTime,
-			}
-
 			if err := tx.WithContext(
 				ctx,
 			).Model(
-				&run,
+				&models.Run{},
 			).Where(
 				"experiment_id = ?", experiment.ID,
-			).Updates(&run).Error; err != nil {
+			).Updates(&models.Run{
+				LifecycleStage: experiment.LifecycleStage,
+				DeletedTime:    experiment.LastUpdateTime,
+			}).Error; err != nil {
 				return eris.Wrapf(err, "error updating existing runs with experiment id: %d", *experiment.ID)
 			}
 		}
 		return nil
-	}); err != nil {
-		return err
-	}
-
-	return nil
+	})
 }
 
-// Delete removes the existing models.Experiment from the db.
+// Delete deletes existing experiment.
 func (r ExperimentRepository) Delete(ctx context.Context, experiment *models.Experiment) error {
-	return r.DeleteBatch(ctx, []*int32{experiment.ID})
-}
-
-// DeleteBatch removes existing []models.Experiment in batch from the db.
-func (r ExperimentRepository) DeleteBatch(ctx context.Context, ids []*int32) error {
 	if err := r.db.Transaction(func(tx *gorm.DB) error {
-		// finding all the runs
+		// finding all the related runs.
 		var minRowNum sql.NullInt64
 		if err := tx.Model(
 			&models.Run{},
 		).Where(
-			"experiment_id IN (?)", ids,
-		).Pluck("MIN(row_num)", &minRowNum).Error; err != nil {
+			"experiment_id  = ?", *experiment.ID,
+		).Pluck(
+			"MIN(row_num)", &minRowNum,
+		).Error; err != nil {
 			return err
 		}
 
-		experiments := make([]models.Experiment, 0, len(ids))
+		// delete current experiment.
 		if err := tx.Clauses(
 			clause.Returning{Columns: []clause.Column{{Name: "experiment_id"}}},
 		).Where(
-			"experiment_id IN ?", ids,
-		).Delete(&experiments).Error; err != nil {
-			return eris.Wrapf(err, "error deleting existing experiments with ids: %d", ids)
-		}
-
-		// verify deletion
-		if len(experiments) != len(ids) {
-			return eris.New("count of deleted experiments does not match length of ids input (invalid experiment ID?)")
+			experiment,
+		).Delete(
+			&models.Experiment{},
+		).Error; err != nil {
+			return eris.Wrapf(err, "error deleting existing experiment with id: %d", *experiment.ID)
 		}
 
-		// renumbering the remainder runs
+		// renumbering the remainder runs.
 		if minRowNum.Valid {
-			runRepo := NewRunRepository(tx)
-			if err := runRepo.renumberRows(tx, models.RowNum(minRowNum.Int64)); err != nil {
-				return eris.Wrapf(err, "error renumbering runs.row_num")
+			if models.RowNum(minRowNum.Int64) < models.RowNum(0) {
```

Comment:
why move the renumberRows code here? I think preferable to keep in its own function.

Reply:
maybe experiment service handle delegation to the two repositories?

Reply:
i think we should do this now -- the renumbering is kind of thorny and we shouldn't have it in two places